nucliadb_utils/nucliadb_utils/nats/demux.py

```diff
@@ -0,0 +1,310 @@
+        self,
+        *,
+        processor: DemuxProcessor,
+        queue_klass: Type[asyncio.Queue] = asyncio.Queue,
+    ):
+        self.processor = processor
+        self.queue_klass = queue_klass
+        self.workers: dict[str, "DemuxWorker"] = {}
+        self.working_on: set[str] = set()
+
+    def handle_message_nowait(self, msg: Msg):
+        """Eventually process `msg` asynchronously and ack/nak it when finished."""
+        subject = msg.subject
+        reply = msg.reply
+        seqid = int(msg.reply.split(".")[5])
+        logger.debug(
+            "Message will be soon demultiplexed and processed",
+            extra={"subject": subject, "reply": reply, "seqid": seqid},
+        )
+        asyncio.create_task(self.safe_demux(msg))
+
+    async def safe_demux(self, msg: Msg):
+        try:
+            await self.demux(msg)
+        except Exception as exc:
+            subject = msg.subject
+            reply = msg.reply
+            seqid = int(msg.reply.split(".")[5])
+
+            event_id = capture_exception(exc)
+            logger.error(
+                "Error while concurrently handling a NATS message. "
+                f"Check sentry for more details. Event id: {event_id}",
+                extra={
+                    "subject": subject,
+                    "reply": reply,
+                    "seqid": seqid,
+                },
+                exc_info=exc,
+            )
+
+            await msg.nak()
+            raise exc
+
+    async def demux(self, msg: Msg):
+        async with MessageProgressUpdater(
+            msg, nats_consumer_settings.nats_ack_wait * 0.66
+        ):
+            subject = msg.subject
+            reply = msg.reply
+            seqid = int(msg.reply.split(".")[5])
+
+            split, work = self.processor.splitter(msg)
+            logger.info(
+                f"Message demultiplexed to split '{split}'",
+                extra={"subject": subject, "reply": reply, "seqid": seqid},
+            )
+
+            worker, created = self.get_or_create_worker(split)
+
+            seqid = int(msg.reply.split(".")[5])
+            work_id = f"{split}:{seqid}"
+            work_unit = WorkUnit(work_id, work)
+
+            if work_id in self.working_on:
+                logger.warning(
+                    f"Trying to process again an already received NATS message. Skipping (workid={work_id})"
+                )
+                return
+
+            await worker.add_work(work_unit)
+            self.working_on.add(work_id)
+
+            if created or (not created and not worker.working):
+                # The second condition should never happen, the worker task finished
+                # between us getting the worker and adding work. We'll create a new
+                # task for it
+                self.start_worker(split)
+
+            loop = asyncio.get_event_loop()
+            wait_condition = loop.create_future()
+
+            work_unit_done.add_listener(
```

Comment:
isn't this a memory leak? We never remove the listener.

Could this also be moved to a global class without a partial and then pass in the NatsDemultiplexer instance instead of passing in partials and having dynamic listeners?

Reply:
Oh, it is... Good catch!

I'm not sure I follow everything. We can discuss it later